sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala

```diff
@@ -221,8 +221,18 @@ class V2ExpressionBuilder(e: Expression, isPredicate: Boolean = false) extends L
     case _: BitwiseNot => generateExpressionWithName("~", expr, isPredicate)
     case caseWhen @ CaseWhen(branches, elseValue) =>
       val conditions = branches.map(_._1).flatMap(generateExpression(_, true))
-      val values = branches.map(_._2).flatMap(generateExpression(_))
-      val elseExprOpt = elseValue.flatMap(generateExpression(_))
+      val values = branches.map(_._2).flatMap(child =>
+        generateExpression(
+          child,
+          isPredicate && child.dataType.isInstanceOf[BooleanType]
```

Comment:
do we need `child.dataType.isInstanceOf[BooleanType]`? I think `generateExpression` only use v2 Predicate if the expression data type is boolean

Reply:
idk seems safer this way. I've seen somewhere in this file a similar check and made me wonder if we can have a Predicate that returns something else.

I would keep it like this just in case, better safe then sorry.

Reply:
I agree with @cloud-fan, if there is a design problem of this class (not being safe against data type check) lets fix it at class level and not per expression level

Reply:
how should we do that?

Reply:
to ensure predicates are booleans always and that only predicates are booleans?